php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+    		        ' WHERE TargetLevel = %3$u '.
+                    '   AND DemonstratedLevel > 0 '.
+                    '   AND Override = false '.
+                    '   AND demoskill.SkillID IN (%4$s) '.
+                    '   AND demo.StudentID = %5$u '.
+                    ' GROUP BY SkillID '.
+    		'  ) ds2 '.
+    		'    ON ds2.SkillID = ds.SkillID '.
+    		'   AND ds2.Created = ds.Created '.
+    		' WHERE ds.SkillID IN (%4$s) '.
+    		'   AND ds.TargetLevel = %3$u '.
+    		'   AND ds.DemonstratedLevel > 0 '.
+    		'   AND d.StudentID = %5$u '.
+    		'   AND ds.Override = false ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                $this->Level,
+                join(', ', $validSkillIds),
+                $this->StudentID
+            ]
+        );
+
+        $this->BaselineRating = array_sum($demonstrationSkillRatings) / count($validSkills);
+
+        if ($autoSave === true) {
+            $this->save();
+        }
+    }
+
+    private $demonstrationData;
+    public function getDemonstrationData()
+    {
+        if (!empty($this->demonstrationData)) {
+            return $this->demonstrationData;
+        }
+
+        try {
+            $data = DB::arrayTable(
+                'SkillID',
+                'SELECT '.
+                '    DemonstrationSkill.*, '.
+                '    Demonstration.Demonstrated AS DemonstrationDate '.
+                '  FROM `%s` DemonstrationSkill '.
+                '  JOIN (SELECT ID, Demonstrated FROM `%s` WHERE StudentID = %u) Demonstration '.
+                '    ON Demonstration.ID = DemonstrationSkill.DemonstrationID '.
+                ' WHERE DemonstrationSkill.SkillID IN (%s) '.
+                '   AND DemonstrationSkill.TargetLevel = %u '.
+                ' ORDER BY SkillID, DemonstrationDate',
+                [
+                    DemonstrationSkill::$tableName,
+                    Demonstration::$tableName,
+                    $this->StudentID,
+                    join(', ', $this->Competency->getSkillIds()),
+                    $this->Level
+                ]
+            );
+
+            return $this->demonstrationData = $data;
+        } catch (\Exception $e) {
+            throw $e;
+            return [];
+        }
+    }
+
+    private $effectiveDemonstrationsData;
+    public function getEffectiveDemonstrationsData()
+    {
+        if (!empty($this->effectiveDemonstrationsData)) {
+            return $this->effectiveDemonstrationsData;
+        }
+
+        $demonstrationsData = $this->getDemonstrationData();
+
+        foreach ($demonstrationsData as $skillId => &$demonstrationData) {
+            $sort = [];
+            foreach ($demonstrationData as $key => $row) {
+                $sort[$key]  = $row['DemonstratedLevel'];
+            }
+            // sort demonstrations by highest level
+            array_multisort($sort, SORT_DESC, $demonstrationData);
```

Comment:
I'm having trouble following what's going on here with `$sort` and `array_multisort`

You just need
```php
protected static function sortEffectiveDemonstrations($a, $b) {
    if ($a['DemonstratedLevel'] == $b['DemonstratedLevel']) {
        return 0;
    }
    return ($a['DemonstratedLevel'] < $b['DemonstratedLevel']) ? -1 : 1;
}
// ...
    foreach ($demonstrationsData as $skillId => &$demonstrationData) {
        uasort($demonstrationData, [__CLASS__,  'sortEffectiveDemonstrations']);
    }
```